### PR TITLE
core: remount sysfs when network and mount namespace are enabled

### LINKED ISSUE #26422: RootImage= with NetworkNamespacePath= does not mount /sys in the correct namespace
### systemd version the issue has been seen with

252

### Used distribution

Debian 11

### Linux kernel version used

6.0.0-0.deb11.6-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

```
root@ciliumlb01 (prod!):~# systemd-run -t --collect -p RootImage=/tmp/image -p PrivateNetwork=yes ls -1 /sys/class/net
lo
```

### Unexpected behaviour you saw

```
root@ciliumlb01 (prod!):~# systemd-run -t --collect -p RootImage=/tmp/image -p PrivateNetwork=yes ls -1 /sys/class/net
eth0
lo
```

### Steps to reproduce the problem

Run the command in the unexpected behaviour. The problem here is that MountAPIVFS= does not change namespace before setting up mounts. According to sysfs documentation the namespace tag on /sys is set to the one mounting.

This affects NetworkNamespacePath= as well.

Workaround:
```
# systemd-run -t --collect -p RootImage=/tmp/image -p PrivateNetwork=yes -p MountAPIVFS=off /bin/bash -c 'mount sysfs -t sysfs /sys; ls -1 /sys/class/net'
Running as unit: run-u100.service
Press ^] three times within 1s to disconnect TTY.
lo
```

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #26422.

### CODE REVIEW

File: src/shared/mount-util.c [hunk 1]
```diff
@@ -1167,6 +1168,161 @@ int remount_idmap(
...
+
+        return 0;
+}
+
+int remount_sysfs(const char *path) {
+        SubMount *mounts;
+        size_t n;
+        int r;
+
+        assert(path);
```
Comment (poettering) [inline src/shared/mount-util.c]: Maybe add a quick comment what this actually does, i.e. why we need something so involved here. It's not obvious to the random reader I think.

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -1663,6 +1663,10 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
         not available), and the unit should be written in a way that does not solely rely on this setting for
         security.</para>
 
+        <para>When this option is enabled, <varname>PrivateMounts=</varname> is also enabled unless it is
+        explicitly disabled, and <filename>/sys</filename> will be remounted to make it tagged with the new
```
Comment (poettering) [inline man/systemd.exec.xml]: "make it tagged with" → "associate it with"?

File: man/systemd.exec.xml [hunk 3]
```diff
@@ -1684,6 +1688,10 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
         <varname>NetworkNamespacePath=</varname> configured, as otherwise the network namespace of those
         units is reused.</para>
 
+        <para>When this option is enabled, <varname>PrivateMounts=</varname> is also enabled unless it is
+        explicitly disabled, and <filename>/sys</filename> will be remounted to make it tagged with the new
+        network namespace.</para>
```
Comment (poettering) [inline man/systemd.exec.xml:1693]: ditto

File: man/systemd.exec.xml [hunk 4]
```diff
@@ -1684,6 +1688,10 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
         <varname>NetworkNamespacePath=</varname> configured, as otherwise the network namespace of those
         units is reused.</para>
 
+        <para>When this option is enabled, <varname>PrivateMounts=</varname> is also enabled unless it is
```
Comment (poettering) [inline man/systemd.exec.xml]: "also enabled" → "implied"?

(similare above)

**Comment by bluca** [comment]
```
../src/shared/mount-util.c: In function ‘remount_sysfs’:
../src/shared/mount-util.c:1286:30: error: ‘n’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 1286 |         for (size_t i = 0; i < n; i++) {
      |                            ~~^~~
../src/shared/mount-util.c:1302:16: note: ‘n’ was declared here
 1302 |         size_t n;
      |                ^
../src/shared/mount-util.c:1301:19: error: ‘mounts’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 1301 |         SubMount *mounts;

```

**Comment by yuwata** [comment]
@poettering and @bluca Thank you for your comments. All suggestions are addressed. Moreover,
- also introduced `exec_needs_ipc_namespace()` helper function,
- fixed missing condition for creating ExecRuntime object,
- made `remount_sysfs()` more generic, 
- and more tests.

PTAL.

File: src/shared/mount-util.c [hunk 5]
```diff
@@ -1167,6 +1168,182 @@ int remount_idmap(
...
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse /proc/self/mountinfo: %m");
+
+        for (;;) {
+                _cleanup_close_ int mount_fd = -EBADF;
+                _cleanup_free_ char *p = NULL;
+                struct libmnt_fs *fs;
+                const char *path;
+
+                r = mnt_table_next_fs(table, iter, &fs);
```
Comment (poettering) [inline src/shared/mount-util.c:1236]: hmm, i guess we could use mnt_table_next_child_fs() here?

File: src/shared/mount-util.c [hunk 6]
```diff
@@ -1167,6 +1168,182 @@ int remount_idmap(
...
+        typesafe_qsort(mounts, n, sub_mount_compare);
+        sub_mount_drop(mounts, n);
+
+        *ret_mounts = TAKE_PTR(mounts);
+        *ret_n_mounts = n;
+        return 0;
+
+fail:
+        assert(r < 0);
+        sub_mount_free_n(mounts, n);
```
Comment (poettering) [inline src/shared/mount-util.c]: candidate for CLEAN_ARRAY() → #26555

Comment (yuwata) [reply src/shared/mount-util.c]: Nice. I will apply it after the PR is merged.

File: src/shared/mount-util.c (hunk 5 above)
Comment (yuwata) [reply src/shared/mount-util.c:1236]: Unfortunately, it cannot. as child mount may be overmounted.

**Comment by dechamps** [comment]
@yuwata It would have been nice to have this be mentioned in [`NEWS`](https://github.com/systemd/systemd/blob/main/NEWS), as `PrivateNetwork=yes` implying `PrivateMounts=yes` (see c2da3bf2376440526e87da6124660d22ac2079d6) is a non-trivial change in behaviour that [caught some people off-guard](https://github.com/systemd/systemd/issues/2741#issuecomment-1661048428) :(

**Pull request merged**